# Import the relevant libraries

In [349]:
import numpy as np
import pandas as pd 

# Load the data

In [351]:
data_preprocessed = pd.read_csv('F:\\schulich\python\\Udemy Data course\\required data files\\Absenteeism_preprocessed.csv')
# let's eyeball the data
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

We expect that half of our features are not so important, but to me it seems that the reason for absence would be the most indicative, maybe the workload is also determinative, since the more busy a person is the less likely that person is to skip the work, and the number of children and pets should also have something to do with absence.
The good thing about regression is that the model itself tells us which features are important and which are not (by comparing their weights)
So we have decided to conduct a logistic regression model.

# Preprocessing based on our model requiements
We will use the logistic regression model, which is a classification method.

## 1- Create the target
we will use a simple approach to deal with targets, we assume two classes: moderately absent and excessively absent and to define which one is which we take the median, if the abesnteeism time is less than median it is considered to be a moderatley absent case and is it is higher so we have the excessively absent case. 

#### Why using median?
1- because it is not affected by outliers

2- we implicitly balance the data sine half of our observations are below and half are above medium (as balancing prevent the model from learning to output only 0s or only 1s)

In [352]:
# let's find the median
data_preprocessed['Absenteeism Time in Hours'].median()

3.0

In [353]:
# Absenteeism Time in Hours <= 3 : moderately absent : return 0
# Absenteeism Time in Hours >= 4 : excessively absent: return 1
# inseted of For loop and IF we can use WHERE method
# np.where(condition, value if true, value if false)
targets = np.where(data_preprocessed['Absenteeism Time in Hours'] <= data_preprocessed['Absenteeism Time in Hours'].median(), 0, 1)
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [354]:
# let's add it to our DataFrame
data_preprocessed['Excessive Absenteeism'] = targets
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0               30          0         2     1                          4   
1               31          0         1     0                          0   
2               31          0         0     0                          2   
3               24          0         2     0                          4   
4               30          0         2     1                          2   

   Excessive Absenteeism  
0                      1  
1                      0  
2                      0  
3                      1  
4                      0

In [355]:
# check for balancing, should be around 50%
targets.sum() / targets.shape[0]

0.45571428571428574

In [356]:
# drop the "Absenteeism time in hour" since we do not need it anymore
# and also save it in another dataframe as checkpoint
data_with_target = data_preprocessed.drop(['Absenteeism Time in Hours'], axis = 1)
data_with_target.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0               30          0         2     1                      1  
1               31          0         1     0                      0  
2               31          0         0     0                      0  
3               24          0         2     0                      1  
4               30          0         2     1                      0

## 2- Select the inputs

In [357]:
# use iloc : DataFrame.iloc[row indeices, column indices]
# we want to select all rows and all columns except 'Exessive Absenteeism'
data_with_target.iloc[: , 0:14] # iloc excludes the ending index: 0:14 = 0,1,2,...,13

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  
0                 30          0         2     1  
1                 31          0         1     0  
2                 31          0         0     0  
3                 24          0         2     0  
4                 30          0         2     1  
..               ...        ...       ...   ...  
695               22          1         2     0  
696               24          0         1     2  
697               25          1         0     0  
698               25          1         0     0  
699               25          0         1     1  

[700 rows x 14 columns]

In [358]:
# if the number of coulmns is too big, sometimes we can'e easily count them, so we can start from the end.
# also in order to change it from temporary to permanently we need to store it in a new dataframe
unscaled_inputs = data_with_target.iloc[: , 0:-1]  # or data_with_target.iloc[: , :-1] or data_with_target.iloc[: , 0:14]

## 3- Standardize Data

In [359]:
# lets import the relevant module
from sklearn.preprocessing import StandardScaler

In [360]:
# create an empty standard scaler object in which there is no information
absenteeism_scaler = StandardScaler()

In [361]:
# put (fit) our input data into this scaler object
absenteeism_scaler.fit(unscaled_inputs)
# we have the information and the mechanism, but the inputs are still unscaled.

StandardScaler(copy=True, with_mean=True, with_std=True)

In [362]:
# in order to apply the scaling mechanism we must use another method called transform
scaled_inputs = absenteeism_scaler.transform(unscaled_inputs)
scaled_inputs

array([[-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         0.88046927,  0.26848661],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.01928035, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.91902997, -0.58968976],
       ...,
       [ 1.73205081, -0.09298136, -0.31448545, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.01928035,  0.26848661]])

In [363]:
scaled_inputs.shape

(700, 14)

## 4- Split the data into train and test and shuffle them
Using train_test_split we can do splitting and shuffling simultaneously 

In [364]:
# import the relevant module
from sklearn.model_selection import train_test_split

In [365]:
X_train, X_test, Y_train, Y_test = train_test_split(scaled_inputs, targets, train_size = 0.8, shuffle = True)

In [366]:
print ('X_train shape =', X_train.shape, '    Y_train shape = ', Y_train.shape)
print ('X_test shape =', X_test.shape, '     Y_test shape = ', Y_test.shape)

X_train shape = (560, 14)     Y_train shape =  (560,)
X_test shape = (140, 14)      Y_test shape =  (140,)


In [367]:
# as shuffling is a random funnction, everytime we reryn the programs we get a new data, because shuffling
# gives a new random order of data. so all random functions in python have a random_state parameter, by which 
# we can avoid different random orders every time. 
X_train, X_test, Y_train, Y_test = train_test_split(scaled_inputs, targets, train_size = 0.8, shuffle = True, random_state = 20)

# Build the model
between statsmodel and sklearn we choose to use sklearn since statsmodel would have a lot of warnings and sklearn is much more stable and warning free

## 1- import the relevant modules

In [368]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

## 2- Train the model

In [369]:
# declare a logistic regression object
reg = LogisticRegression()

In [370]:
# fit the regression
reg.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [372]:
# let's evaluate the model accuracy: almost 80% accuracy
reg.score(X_train, Y_train)

0.7839285714285714

In [373]:
# lets calculate the accuracy manualy
model_outputs = reg.predict(X_train)
model_outputs

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,

In [374]:
Y_train

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,

In [375]:
# it is hard to compare MODEL_OUTPUT with Y_train (Target_train) with naked eyes
# lets make it easier to compare by just looking at data
Comparison_array = model_outputs == Y_train
Comparison_array

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True, False, False,  True,  True,  True,  True,
       False,  True, False,  True, False, False,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False, False, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,  True,  True, False,  True, False,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
       False,  True, False,  True,  True, False, False, False,  True,
        True,  True,  True,  True,  True,  True,  True, False,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False,  True,  True,  True,
        True,  True,

In [376]:
# lets calculate the number of True entries: 
# as True is supposed to be 1, we can use Sum function to calculate it
np.sum(Comparison_array)

439

In [377]:
# lets calculate the accuracy 
train_accuracy = np.sum(Comparison_array) / model_outputs.shape[0]
train_accuracy

0.7839285714285714

## 3- Find the intercept and coefficients

In [378]:
reg.intercept_

array([-0.22206736])

In [379]:
reg.coef_

array([[ 2.07601767,  0.33504757,  1.56162303,  1.32927434,  0.18793677,
        -0.07062253,  0.70639316, -0.03986811, -0.20089491, -0.00456366,
         0.31933564, -0.135508  ,  0.38172443, -0.3332426 ]])

In [380]:
# the above results are a bit confusing, we want to make it more orgnanized
scaled_inputs.columns.values

AttributeError: 'numpy.ndarray' object has no attribute 'columns'

In [381]:
# from the above we got error saying that: 'numpy.ndarray' object has no attribute 'columns'
# because our scaled_inputs is an ndarray and not a DataFrame 
# bacause although our unscaled_inputs is a DataFrame:
unscaled_inputs

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  
0                 30          0         2     1  
1                 31          0         1     0  
2                 31          0         0     0  
3                 24          0         2     0  
4                 30          0         2     1  
..               ...        ...       ...   ...  
695               22          1         2     0  
696               24          0         1     2  
697               25          1         0     0  
698               25          1         0     0  
699               25          0         1     1  

[700 rows x 14 columns]

In [382]:
# but we used sklearn to standardize it. and sklearn changes the type of DataFrame to ndarray
# because sklearn is not comatible with pandas DataFrame
scaled_inputs

array([[-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         0.88046927,  0.26848661],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.01928035, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.91902997, -0.58968976],
       ...,
       [ 1.73205081, -0.09298136, -0.31448545, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ...,  2.23224237,
        -0.91902997, -0.58968976],
       [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
        -0.01928035,  0.26848661]])

In [383]:
# so in order to get the title of the columns instead of scaled_inputs which is an array
# we can use unscaled_inputs which is a DataFrame
unscaled_inputs.columns.values

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Day of the Week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [384]:
# it is always a good idea to declare a variable called feature name
feature_name = unscaled_inputs.columns.values

In [385]:
# lets create a DataFrame that contains the feature names
summary_table = pd.DataFrame(data= feature_name, columns= ['Feature Name'])
summary_table

Feature Name
0                  Reason_1
1                  Reason_2
2                  Reason_3
3                  Reason_4
4               Month Value
5           Day of the Week
6    Transportation Expense
7          Distance to Work
8                       Age
9   Daily Work Load Average
10          Body Mass Index
11                Education
12                 Children
13                     Pets

In [386]:
# add another column for corresponding coefficients
summary_table['Coefficient'] = np.transpose(reg.coef_) # note that we must transpose the array because by default it is a row and not column
summary_table 

Feature Name  Coefficient
0                  Reason_1     2.076018
1                  Reason_2     0.335048
2                  Reason_3     1.561623
3                  Reason_4     1.329274
4               Month Value     0.187937
5           Day of the Week    -0.070623
6    Transportation Expense     0.706393
7          Distance to Work    -0.039868
8                       Age    -0.200895
9   Daily Work Load Average    -0.004564
10          Body Mass Index     0.319336
11                Education    -0.135508
12                 Children     0.381724
13                     Pets    -0.333243

In [387]:
# add a row for intercept
# first we need to shift all the values one unit down
summary_table.index = summary_table.index + 1 
summary_table

Feature Name  Coefficient
1                  Reason_1     2.076018
2                  Reason_2     0.335048
3                  Reason_3     1.561623
4                  Reason_4     1.329274
5               Month Value     0.187937
6           Day of the Week    -0.070623
7    Transportation Expense     0.706393
8          Distance to Work    -0.039868
9                       Age    -0.200895
10  Daily Work Load Average    -0.004564
11          Body Mass Index     0.319336
12                Education    -0.135508
13                 Children     0.381724
14                     Pets    -0.333243

In [388]:
# # we put the intercept in the row 0
# we need to sort the data by index since 0 is in the last position
# and also change the for of intercept from a list to a number (very important, even if you do not do it now and in the next
# line you fix it you changed the numbers type and when you want to take exp you get error)
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]
summary_table.sort_index()

Feature Name  Coefficient
0                 Intercept    -0.222067
1                  Reason_1     2.076018
2                  Reason_2     0.335048
3                  Reason_3     1.561623
4                  Reason_4     1.329274
5               Month Value     0.187937
6           Day of the Week    -0.070623
7    Transportation Expense     0.706393
8          Distance to Work    -0.039868
9                       Age    -0.200895
10  Daily Work Load Average    -0.004564
11          Body Mass Index     0.319336
12                Education    -0.135508
13                 Children     0.381724
14                     Pets    -0.333243

### interpreting the coefficinets

note that as all variables have been standardized, we have standadized coefficents as well, meaning that we can compare these weights to each other. the bigger the absolute value of the weight, the more important with bigger influenct its corrsponding feature is. 


In [389]:
# lets add the absolute values of the weights
summary_table['|Coefficient|'] = np.absolute(summary_table['Coefficient'])
summary_table

Feature Name  Coefficient  |Coefficient|
1                  Reason_1     2.076018       2.076018
2                  Reason_2     0.335048       0.335048
3                  Reason_3     1.561623       1.561623
4                  Reason_4     1.329274       1.329274
5               Month Value     0.187937       0.187937
6           Day of the Week    -0.070623       0.070623
7    Transportation Expense     0.706393       0.706393
8          Distance to Work    -0.039868       0.039868
9                       Age    -0.200895       0.200895
10  Daily Work Load Average    -0.004564       0.004564
11          Body Mass Index     0.319336       0.319336
12                Education    -0.135508       0.135508
13                 Children     0.381724       0.381724
14                     Pets    -0.333243       0.333243
0                 Intercept    -0.222067       0.222067

In [390]:
# so let's sort the data base on their importance
summary_table_sorted_based_on_importance = summary_table.sort_values('|Coefficient|', ascending=False)
summary_table_sorted_based_on_importance

Feature Name  Coefficient  |Coefficient|
1                  Reason_1     2.076018       2.076018
3                  Reason_3     1.561623       1.561623
4                  Reason_4     1.329274       1.329274
7    Transportation Expense     0.706393       0.706393
13                 Children     0.381724       0.381724
2                  Reason_2     0.335048       0.335048
14                     Pets    -0.333243       0.333243
11          Body Mass Index     0.319336       0.319336
0                 Intercept    -0.222067       0.222067
9                       Age    -0.200895       0.200895
5               Month Value     0.187937       0.187937
12                Education    -0.135508       0.135508
6           Day of the Week    -0.070623       0.070623
8          Distance to Work    -0.039868       0.039868
10  Daily Work Load Average    -0.004564       0.004564

In [391]:
# lets add odds to our table
summary_table_sorted_based_on_importance['Odds_ratio'] = np.exp(summary_table.Coefficient)
summary_table_sorted_based_on_importance

Feature Name  Coefficient  |Coefficient|  Odds_ratio
1                  Reason_1     2.076018       2.076018    7.972656
3                  Reason_3     1.561623       1.561623    4.766551
4                  Reason_4     1.329274       1.329274    3.778301
7    Transportation Expense     0.706393       0.706393    2.026668
13                 Children     0.381724       0.381724    1.464808
2                  Reason_2     0.335048       0.335048    1.398007
14                     Pets    -0.333243       0.333243    0.716596
11          Body Mass Index     0.319336       0.319336    1.376213
0                 Intercept    -0.222067       0.222067    0.800861
9                       Age    -0.200895       0.200895    0.817998
5               Month Value     0.187937       0.187937    1.206757
12                Education    -0.135508       0.135508    0.873272
6           Day of the Week    -0.070623       0.070623    0.931814
8          Distance to Work    -0.039868       0.039868    0.960916
10  Daily Work Load Average    -0.004564       0.004564    0.995447

### How to interepret the odds_ratio?
As you see from the above table if coefficient is around zero it odds_ratio is around 1.
##### odds_ratio = for one unit change in the standardized feature, the odds increase by a multiple equal to the odds ratio. (odds_ratio = 1 => no change)
in the below formula if bk = 0 then e^bk = e^0 = 1 , so change in X (X2-X1) has no effect in Odds : (Odds2 - Odds1 = 0 )

$ log{\frac{P(X)}{1-P(X)}} = b_kX + b_0 =>$ 
$ log{\frac{odds_2}{odds_1}} = {{b_k}*{(X_2-X_1)}} =>$
$ \frac{odds_2}{odds_1} = e^{{b_k}*{(X_2-X_1)}} =>$
$ odds_2 = e^{{b_k}*{(X_2-X_1)}}*odds_1 $

if odds = 5:1 and odds_ratio (e^bk) = 2 then if x increases by 1 unit it odds inceases by 2 units : New odds = 10:1

for example the odds_ratio for transportation expense is 2, so if the tranportation expense increases by 1 unit (for standardized inputs 1 unit is kind of vague) , its odds for being excessively absent increases by 2 units. 
Features 6, 8, 10 have the coefficents of almost 0 and the odds ratio of 1, so they are not so important on determining the absence status of employee so we consider dropping them later on

#### What about our dummy features? 
when we were creating the dummies, the one that we droped was reason_0, which was the case that the person was absent without any particular reason, therefore the base model is when there is no reason, from reason_1, 2, 3, and 4 as their weights are higher than one, it means that when an employee has a reason, their chance of being excessively absent is higher compared with the situation where they do not declare any specific reason. A good question would be how much bigger of a chance, can we say for example for reaon_1 since its odds_ratio is 7.92 so its odds is 8 units higher? THE ANSWER IS NO.
Because we made a mistake when we standardize our dummies along with numerical inputs. which was a bad practice, because when we standardize the dummies we lost the whole interpretability of the dummies. if we left them 0 and 1s, we could have said for example if the reason for absent is Reason_1, it is 7.92 times more likely that a person will be excessively absent compared to no reason given. by standardizing the dummies the predictive power of the model is still valid, and also its accuracy is higher, but if we use unstandardized dummies, the accuracy may decreases, but it would be more intepretable and we can more easily compare the reaons for absence. so we need to correct the model in a way that our dummies remain untoched and unstandardized. (education is also dummy)

## Build the model agian this time with 0 and 1 dummies (unstandardized dummies)

In [417]:
data_with_target.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0               30          0         2     1                      1  
1               31          0         1     0                      0  
2               31          0         0     0                      0  
3               24          0         2     0                      1  
4               30          0         2     1                      0

#### Backward Elimination: rerun the model this time without unimporatant features
unimportant feature are the features in the previuos model that are close to zero coefficients (weights)

In [418]:
# as in the above analysis we found that the last three features are not important. we need to drop them
data_with_target_without_unimportant_features = data_with_target.drop(['Day of the Week','Distance to Work', 'Daily Work Load Average'], axis=1)
unscaled_inputs = data_with_target_without_unimportant_features.iloc[:, :-1]
unscaled_inputs


Reason_1  Reason_2  Reason_3  Reason_4  Month Value  \
0           0         0         0         1            7   
1           0         0         0         0            7   
2           0         0         0         1            7   
3           1         0         0         0            7   
4           0         0         0         1            7   
..        ...       ...       ...       ...          ...   
695         1         0         0         0            5   
696         1         0         0         0            5   
697         1         0         0         0            5   
698         0         0         0         1            5   
699         0         0         0         1            5   

     Transportation Expense  Age  Body Mass Index  Education  Children  Pets  
0                       289   33               30          0         2     1  
1                       118   50               31          0         1     0  
2                       179   38               31          0         0     0  
3                       279   39               24          0         2     0  
4                       289   33               30          0         2     1  
..                      ...  ...              ...        ...       ...   ...  
695                     179   40               22          1         2     0  
696                     225   28               24          0         1     2  
697                     330   28               25          1         0     0  
698                     235   32               25          1         0     0  
699                     291   40               25          0         1     1  

[700 rows x 11 columns]

In [419]:
# insead of the object StandardScaler we create our own CustomScaler object

In [420]:
# import the relevant modules
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler

# create the class which works similar to StandardScaler with only one difference
# StandardScaler standardizes all the inputs but our CustomScaler only standardzie the 
# the only columns that we choose. 
   
    # create the Custom Scaler class

class CustomScaler(BaseEstimator,TransformerMixin): 
    
    # init or what information we need to declare a CustomScaler object
    # and what is calculated/declared as we do
    
    def __init__(self,columns,copy=True,with_mean=True,with_std=True):
        
        # scaler is nothing but a Standard Scaler object
        self.scaler = StandardScaler(copy,with_mean,with_std)
        # with some columns 'twist'
        self.columns = columns
        self.mean_ = None
        self.var_ = None
        
    
    # the fit method, which, again based on StandardScale
    
    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns])
        self.var_ = np.var(X[self.columns])
        return self
    
    # the transform method which does the actual scaling

    def transform(self, X, y=None, copy=None):
        
        # record the initial order of the columns
        init_col_order = X.columns
        
        # scale all features that you chose when creating the instance of the class
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)
        
        # declare a variable containing all information that was not scaled
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]
        
        # return a data frame which contains all scaled features and all 'not scaled' features
        # use the original order (that you recorded in the beginning)
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]

In [402]:
unscaled_inputs.columns.values

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Transportation Expense', 'Age', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [422]:
# we will create a new variable called columns to scale ommiting dummy variables for the above list
columns_to_scale = ['Month Value', 'Transportation Expense', 'Age', 'Body Mass Index', 'Children', 'Pets']

In [423]:
# declare the absenteeism scaler
absenteeism_scaler = CustomScaler(columns_to_scale)

In [424]:
# put the data into our scaler
absenteeism_scaler.fit(unscaled_inputs)

C:\Users\sasan\anaconda3\lib\site-packages\sklearn\base.py:197: FutureWarning: From version 0.24, get_params will raise an AttributeError if a parameter cannot be retrieved as an instance attribute. Previously it would return None.
  FutureWarning)


CustomScaler(columns=['Month Value', 'Transportation Expense', 'Age',
                      'Body Mass Index', 'Children', 'Pets'],
             copy=None, with_mean=None, with_std=None)

In [425]:
scaled_inputs = absenteeism_scaler.transform(unscaled_inputs)

In [426]:
scaled_inputs

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  \
0           0         0         0         1     0.182726   
1           0         0         0         0     0.182726   
2           0         0         0         1     0.182726   
3           1         0         0         0     0.182726   
4           0         0         0         1     0.182726   
..        ...       ...       ...       ...          ...   
695         1         0         0         0    -0.388293   
696         1         0         0         0    -0.388293   
697         1         0         0         0    -0.388293   
698         0         0         0         1    -0.388293   
699         0         0         0         1    -0.388293   

     Transportation Expense       Age  Body Mass Index  Education  Children  \
0                  1.005844 -0.536062         0.767431          0  0.880469   
1                 -1.574681  2.130803         1.002633          0 -0.019280   
2                 -0.654143  0.248310         1.002633          0 -0.919030   
3                  0.854936  0.405184        -0.643782          0  0.880469   
4                  1.005844 -0.536062         0.767431          0  0.880469   
..                      ...       ...              ...        ...       ...   
695               -0.654143  0.562059        -1.114186          1  0.880469   
696                0.040034 -1.320435        -0.643782          0 -0.019280   
697                1.624567 -1.320435        -0.408580          1 -0.919030   
698                0.190942 -0.692937        -0.408580          1 -0.919030   
699                1.036026  0.562059        -0.408580          0 -0.019280   

         Pets  
0    0.268487  
1   -0.589690  
2   -0.589690  
3   -0.589690  
4    0.268487  
..        ...  
695 -0.589690  
696  1.126663  
697 -0.589690  
698 -0.589690  
699  0.268487  

[700 rows x 11 columns]

In [427]:
# import the relevant module
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(scaled_inputs, targets, train_size = 0.8, shuffle = True, random_state = 20)
# declare a logistic regression object
reg = LogisticRegression()
# fit the regression
reg.fit(X_train, Y_train)
# let's evaluate the model accuracy: almost 80% accuracy
reg.score(X_train, Y_train)
# as you see the accuracy has decreased a little bit

0.7732142857142857

In [428]:
reg.intercept_

array([-1.6474549])

In [430]:
reg.coef_

array([[ 2.80019733,  0.95188356,  3.11555338,  0.83900082,  0.1589299 ,
         0.60528415, -0.16989096,  0.27981088, -0.21053312,  0.34826214,
        -0.27739602]])

In [431]:
feature_name = unscaled_inputs.columns.values
# lets create a DataFrame that contains the feature names
summary_table = pd.DataFrame(data= feature_name, columns= ['Feature Name'])
summary_table

Feature Name
0                 Reason_1
1                 Reason_2
2                 Reason_3
3                 Reason_4
4              Month Value
5   Transportation Expense
6                      Age
7          Body Mass Index
8                Education
9                 Children
10                    Pets

In [432]:
# add another column for corresponding coefficients
summary_table['Coefficient'] = np.transpose(reg.coef_) # note that we must transpose the array because by default it is a row and not column
summary_table

Feature Name  Coefficient
0                 Reason_1     2.800197
1                 Reason_2     0.951884
2                 Reason_3     3.115553
3                 Reason_4     0.839001
4              Month Value     0.158930
5   Transportation Expense     0.605284
6                      Age    -0.169891
7          Body Mass Index     0.279811
8                Education    -0.210533
9                 Children     0.348262
10                    Pets    -0.277396

In [433]:
summary_table.index = summary_table.index + 1 
summary_table

Feature Name  Coefficient
1                 Reason_1     2.800197
2                 Reason_2     0.951884
3                 Reason_3     3.115553
4                 Reason_4     0.839001
5              Month Value     0.158930
6   Transportation Expense     0.605284
7                      Age    -0.169891
8          Body Mass Index     0.279811
9                Education    -0.210533
10                Children     0.348262
11                    Pets    -0.277396

In [434]:
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]
summary_table.sort_index()

Feature Name  Coefficient
0                Intercept    -1.647455
1                 Reason_1     2.800197
2                 Reason_2     0.951884
3                 Reason_3     3.115553
4                 Reason_4     0.839001
5              Month Value     0.158930
6   Transportation Expense     0.605284
7                      Age    -0.169891
8          Body Mass Index     0.279811
9                Education    -0.210533
10                Children     0.348262
11                    Pets    -0.277396

In [435]:
summary_table['|Coefficient|'] = np.absolute(summary_table['Coefficient'])
summary_table

Feature Name  Coefficient  |Coefficient|
1                 Reason_1     2.800197       2.800197
2                 Reason_2     0.951884       0.951884
3                 Reason_3     3.115553       3.115553
4                 Reason_4     0.839001       0.839001
5              Month Value     0.158930       0.158930
6   Transportation Expense     0.605284       0.605284
7                      Age    -0.169891       0.169891
8          Body Mass Index     0.279811       0.279811
9                Education    -0.210533       0.210533
10                Children     0.348262       0.348262
11                    Pets    -0.277396       0.277396
0                Intercept    -1.647455       1.647455

In [436]:
summary_table_sorted_based_on_importance = summary_table.sort_values('|Coefficient|', ascending=False)
summary_table_sorted_based_on_importance

Feature Name  Coefficient  |Coefficient|
3                 Reason_3     3.115553       3.115553
1                 Reason_1     2.800197       2.800197
0                Intercept    -1.647455       1.647455
2                 Reason_2     0.951884       0.951884
4                 Reason_4     0.839001       0.839001
6   Transportation Expense     0.605284       0.605284
10                Children     0.348262       0.348262
8          Body Mass Index     0.279811       0.279811
11                    Pets    -0.277396       0.277396
9                Education    -0.210533       0.210533
7                      Age    -0.169891       0.169891
5              Month Value     0.158930       0.158930

In [437]:
summary_table_sorted_based_on_importance['Odds_ratio'] = np.exp(summary_table.Coefficient)
summary_table_sorted_based_on_importance

Feature Name  Coefficient  |Coefficient|  Odds_ratio
3                 Reason_3     3.115553       3.115553   22.545903
1                 Reason_1     2.800197       2.800197   16.447892
0                Intercept    -1.647455       1.647455    0.192539
2                 Reason_2     0.951884       0.951884    2.590585
4                 Reason_4     0.839001       0.839001    2.314054
6   Transportation Expense     0.605284       0.605284    1.831773
10                Children     0.348262       0.348262    1.416604
8          Body Mass Index     0.279811       0.279811    1.322880
11                    Pets    -0.277396       0.277396    0.757754
9                Education    -0.210533       0.210533    0.810152
7                      Age    -0.169891       0.169891    0.843757
5              Month Value     0.158930       0.158930    1.172256

In [453]:
# we want to move the intercept to the bottom of our table to make comparison between features easier
summary_table_without_intercept = summary_table_sorted_based_on_importance.drop([0], axis=0)
final_summary_table = summary_table_without_intercept.append(summary_table_sorted_based_on_importance.iloc[2, :])
final_summary_table

Feature Name  Coefficient  |Coefficient|  Odds_ratio
3                 Reason_3     3.115553       3.115553   22.545903
1                 Reason_1     2.800197       2.800197   16.447892
2                 Reason_2     0.951884       0.951884    2.590585
4                 Reason_4     0.839001       0.839001    2.314054
6   Transportation Expense     0.605284       0.605284    1.831773
10                Children     0.348262       0.348262    1.416604
8          Body Mass Index     0.279811       0.279811    1.322880
11                    Pets    -0.277396       0.277396    0.757754
9                Education    -0.210533       0.210533    0.810152
7                      Age    -0.169891       0.169891    0.843757
5              Month Value     0.158930       0.158930    1.172256
0                Intercept    -1.647455       1.647455    0.192539

#### Now we can easily interepret dummy variables:
as you see after changing dummies from standardized to the initial 0 and 1 values and also dropping unimportant features, the weights have changed significantly, 
lets remind the reasons: 0: no reason, 1: various diseases, 2: pregnancy, 3: poisoning, 4: light diseases. the most crucial reason for excessive absence is poisoning, which makes sense as it is something unpredictable. for example the odds_ratio of 22.54 means that the odds of being excessivley absent for someone who has been poinsoned is 22.54 times higher than somebody who is absent with no reason. reason 1 which regards to light diseases is in the second place, which is normal. pregnancy is also important but is it is something managable and does not affect the ability to work ans can't cause absenteeism so much. 

#### How to interepret the standardized numerical coeffiicients:
as we menthoined standardization leads increase to accuracy and prediction power at the cost of losing the interpretability of coefficients. but we can still interpret it in some way. for example the standardized odds_ratio of transportation expense is 1.83. we can say per one unit increase in standdardized transportation expense the odds of being excessively abesnt increases by 1.82 times. so 1 unit increase in standardized value is eaqual to one standard deviation increase in unstandardized values. 
what about pets, its weight is negative and its ratio is below 1. so it means that one increase in the number of pets leads to reduce the odds ob being excessively absent by 75%. which we did not expect. but one explanation would be if you have too many pets you do not take care of all of them by yourself and you have someone who is doing thayt for you. 

# Test the model

## 1- calculate the test accuracy and compare it with training accuracy

In [457]:
reg.score(X_test, Y_test)
# almost all the time the test accuarcy is less than train accuracy, if it is highry you either got lucky or made a mistake.
# often 10% to 20% lower than training accuracy, meaning that our model is overfitted,
# but in this example 75% is good 
# 0.75 < 0.77

0.75

## 2- Predict the probability of being absent
Instead of getting 0 and one we can get the probability of being excessively absent which is more useful in our analysis

In [461]:
predicted_probability = reg.predict_proba(X_test)
predicted_probability
# the first columns it the probability of being 0 and the second is the probability of being 1.

array([[0.71340413, 0.28659587],
       [0.58724228, 0.41275772],
       [0.44020821, 0.55979179],
       [0.78159464, 0.21840536],
       [0.08410854, 0.91589146],
       [0.33487603, 0.66512397],
       [0.29984576, 0.70015424],
       [0.13103971, 0.86896029],
       [0.78625404, 0.21374596],
       [0.74903632, 0.25096368],
       [0.49397598, 0.50602402],
       [0.22484913, 0.77515087],
       [0.07129151, 0.92870849],
       [0.73178133, 0.26821867],
       [0.30934135, 0.69065865],
       [0.5471671 , 0.4528329 ],
       [0.55052275, 0.44947725],
       [0.5392707 , 0.4607293 ],
       [0.40201117, 0.59798883],
       [0.05361575, 0.94638425],
       [0.7003009 , 0.2996991 ],
       [0.78159464, 0.21840536],
       [0.42037128, 0.57962872],
       [0.42037128, 0.57962872],
       [0.24783565, 0.75216435],
       [0.74566259, 0.25433741],
       [0.51017274, 0.48982726],
       [0.85690195, 0.14309805],
       [0.20349733, 0.79650267],
       [0.78159464, 0.21840536],
       [0.

In [464]:
# if we just want to know the probability of getting excessive absenteeism we just use the second column:
predicted_probability[: , 1]
# in reality regression analysis first calculate these propbabilities in the background,
# the if this probability is <= 0.5 it returns 0 otherwise 1. 

array([0.28659587, 0.41275772, 0.55979179, 0.21840536, 0.91589146,
       0.66512397, 0.70015424, 0.86896029, 0.21374596, 0.25096368,
       0.50602402, 0.77515087, 0.92870849, 0.26821867, 0.69065865,
       0.4528329 , 0.44947725, 0.4607293 , 0.59798883, 0.94638425,
       0.2996991 , 0.21840536, 0.57962872, 0.57962872, 0.75216435,
       0.25433741, 0.48982726, 0.14309805, 0.79650267, 0.21840536,
       0.36956558, 0.67906035, 0.68502567, 0.52868083, 0.21840536,
       0.53506551, 0.22147081, 0.73692105, 0.40498044, 0.60505988,
       0.21075848, 0.45224466, 0.23751292, 0.39833498, 0.82755447,
       0.56797575, 0.69113325, 0.28659587, 0.21935267, 0.2033097 ,
       0.57628256, 0.3294664 , 0.66512397, 0.26949499, 0.83321968,
       0.43491525, 0.88374612, 0.23127072, 0.33415858, 0.34432939,
       0.69909345, 0.65494263, 0.29244941, 0.79200758, 0.20750276,
       0.26840558, 0.08708566, 0.22147081, 0.73245417, 0.30530219,
       0.22147081, 0.29014408, 0.90438191, 0.46061297, 0.60174

# Save the model
when we built our model, we want to use it later with new data, because we the weights and the bias and other parameters such as random_state =20, so instead of using weights and the bias, we can save the whole model which is already trained. and just put the data and get the predicted output for them. (no need to run this whole notebook everytime). if you take a closer look at our model, we know that the object reg contains all the needed information to do the prediction. so saving the model is equivilant to saving the object reg. next time when we want to use the model we just upload the model in a new notebook. 

In [465]:
# there are several ways to save the model, the most general way is to use pickle
# once we want to use it we will unpickle it. 
import pickle


In [468]:
# with open('our model name', 'wb(stands for write byte)') as file:  
#    pickle.dump(object name in the current notebook to be dumped, file)   : pickle.dump = load
with open('model', 'wb') as file:  
    pickle.dump(reg, file)

In [469]:
# we also used another object in this notebook to preprocess data, absenteeism_scaler
# we need to save that too
# if we want to preprocess any new data we use this object. 
with open('scaler', 'wb') as file:  
    pickle.dump(absenteeism_scaler, file)